packages/internals/support/src/entities/component.js

```diff
@@ -1,71 +1,112 @@
-const assert = require('check-types').assert;
+const {assert} = require('check-types');
 const check = require('check-more-types');
 const {omit} = require('lodash');
+const {cloneDeep, get, set} = require('lodash');
 const Collection = require('../collections/collection');
 const FileCollection = require('../collections/file-collection');
 const Entity = require('./entity');
 const File = require('./file');
 
-const isFile = value => value instanceof File;
-check.mixin(isFile, 'file');
-const isCollection = value => value instanceof Collection;
-check.mixin(isCollection, 'collection');
-const isFileCollection = value => value instanceof FileCollection;
-check.mixin(isFileCollection, 'fileCollection');
+const _config = new WeakMap();
+const _data = new WeakMap();
+const _defaults = new WeakMap();
 
-const componentSchema = {
-  path: check.maybe.unemptyString,
-  relative: check.maybe.unemptyString,
-  src: check.file,
-  name: check.maybe.unemptyString,
-  config: check.maybe.object,
-  variants: check.maybe.collection,
-  files: check.maybe.fileCollection
-};
-
-const _src = new WeakMap();
-const _files = new WeakMap();
-const _variants = new WeakMap();
-
-const privateProps = ['src', 'files', 'variants'];
+const privateProps = [
+  'src', 'files', 'variants', 'config',
+  'path', 'relative', 'name'
+];
 
 class Component extends Entity {
   constructor(props = {}) {
     assert(check.schema(componentSchema, props), `Component.constructor: The properties provided do not match the schema of a component [properties-invalid]`, TypeError);
 
     props.name = props.name || props.src.stem;
-    props.config = props.config || {};
 
     super(omit(props, privateProps));
 
-    _src.set(this, props.src);
-    _files.set(this, props.files || new FileCollection());
-    _variants.set(this, props.variants || new Collection());
+    _data.set(this, {});
+    _config.set(this, Object.assign({}, props.config));
 
-    // TODO: Revisit if memory impacted
-    Object.defineProperty(this, 'path', {
+    const core = {
+      src: props.src,
+      files: (props.files || new FileCollection()),
+      variants: (props.variants || new Collection()),
+      name: props.name || props.src.stem
+    };
+    Object.defineProperty(core, 'path', {
       enumerable: true,
-      get: pathGet
+      get: () => {
+        return get(_defaults.get(this), 'src').path;
+      }
     });
-
-    Object.defineProperty(this, 'relative', {
+    Object.defineProperty(core, 'relative', {
```

Comment:
Definitely a lot cleaner defining getters on the core data object like this than having `if (path = 'relative')` checks in there! 👍